Allow subqueries in ANY expressions

mxm opened the pull request:
This enables queries like the following: `Select id from t1 where 1 = ANY(Select id from t2)`

Follow-up PR to #6086 which introduces a `SingleColumnTableType` for all subqueries. This PR removes the limitation that subqueries can only return a single value in ANY or IN clauses. For Subqueries which only allow a single value, the `singleValue` function enforces that only one value is returned.

mikethebeer commented on sql/src/test/java/io/crate/integrationtests/SubSelectIntegrationTest.java:
```diff
@@ -23,6 +23,7 @@
 package io.crate.integrationtests;
 
 import io.crate.operation.projectors.sorting.OrderingByPosition;
+import io.crate.testing.SQLResponse;
```
import not needed

mikethebeer commented on sql/src/main/java/io/crate/analyze/expressions/ExpressionAnalyzer.java:
```diff
@@ -774,9 +852,21 @@ protected Symbol visitSubqueryExpression(SubqueryExpression node, ExpressionAnal
+            // A SubQuery can return more than one row. We don't allow multiple rows,
+            // except in ANY or IN expressions. Thus, we wrap the result into a function
+            // which extracts the first element and checks if there are more than one element.
+            if (!node.isArrayExpression()) {
```
I would recommend to use `node.isArrayExpression() == false` to be consistent. It improves readability

mikethebeer replied:
In general this check makes no sense to me. `visitSubqueryExpression` won't get called anyway if `ANY (array_expression)` is used.

mikethebeer commented on sql/src/main/java/io/crate/analyze/expressions/ExpressionAnalyzer.java:
```diff
@@ -530,14 +601,21 @@ public Symbol visitArrayComparisonExpression(ArrayComparisonExpression node, Exp
 
-            Symbol arraySymbol = process(node.getRight(), context);
+            Expression right = node.getRight();
+            if (right instanceof SubqueryExpression) {
```
IMHO the `infstanceof` is not necessary because the `visitSubqueryExpression` will only be called (by `process`) anyway if the parser detected a `SubqueryExpression`.

dobe commented on CHANGES.txt:
```diff
@@ -27,6 +27,10 @@ Breaking Changes
 Changes
 =======
 
+ - Introduce support for single row-single column subselects in ANY, e.g.::
```
can't we just write: "Added support for subselects as an argument to the ANY function"

bg: it is not true that we only support single row, and multicol will not be implemented i guess.

i would also probably leave out the example, which is btw multirow :-)

dobe commented on core/src/main/java/io/crate/types/SingleColumnTableType.java:
```diff
@@ -0,0 +1,141 @@
+import java.util.Arrays;
+
+/**
+ * A type which represents a table retrieved by a subquery with one column.
```
what about: A type which represents a single column table as a collection containing the value of the first column of each row.

dobe commented on core/src/main/java/io/crate/types/SingleColumnTableType.java:
```diff
@@ -0,0 +1,141 @@
+/**
+ * A type which represents a table retrieved by a subquery with one column.
+ *
+ * Currently only supports a single column but can be extended
```
pls delete, this is anticipated, we probably might do it differently in future

dobe commented on core/src/main/java/io/crate/types/SingleColumnTableType.java:
```diff
@@ -0,0 +1,141 @@
+        return this;
+    }
+
+    @Override
```
pls document explicitly again what this returns

dobe commented on core/src/main/java/io/crate/types/SingleColumnTableType.java:
```diff
@@ -0,0 +1,141 @@
+
+    @Override
+    public boolean isConvertableTo(DataType other) {
+        return this.valueType.isConvertableTo(other) ||
```
this is wrong, it would mean you can convert this collection type to any of its inner types.
pls take a look at the implementation in ArrayType (the null check is also wrong)
It would probably make sense to extract a superclass and reuse methods

dobe commented on sql/src/main/java/io/crate/analyze/expressions/ExpressionAnalyzer.java:
```diff
@@ -774,11 +851,23 @@ protected Symbol visitSubqueryExpression(SubqueryExpression node, ExpressionAnal
+            // A SubQuery can return more than one row. We don't allow multiple rows,
+            // except in ANY or IN expressions. Thus, we wrap the result into a function
+            // which extracts the first element and checks if there are more than one element.
+            if (node.isArrayExpression() == false) {
```
this should not be required, we need to support multiple rows and use the singlecol table type for every subselect that  is a  valueexpression.

mxm commented:
This can be subsumed once #6086 has been merged. Note that this PR was already implementing full support for subqueries in ANY apart from a soft limit that we apply on subqueries at the moment.

mxm commented:
@dobe @mibe I've updated the PR. We have support now for subselects in ANY without `collect_set`. I've removed the soft limit 2 from the planner. For single value expressions, we still use the `single_value` function to extract only one row.

mikethebeer commented:
@mxm Looks gud to me 🥇

dobe commented on sql/src/main/java/io/crate/analyze/expressions/ExpressionAnalyzer.java:
```diff
@@ -600,14 +600,23 @@ public Symbol visitArrayComparisonExpression(ArrayComparisonExpression node, Exp
+            Expression right = node.getRight();
+
+            if (right instanceof SubqueryExpression) {
+                context.registerSubqueryExpression((SubqueryExpression) right);
```
i think it would be more straight forward to apply the single value function in this block after calling process on right, you have the symbol then and know what you need to do. also if applying the single value function you need to apply a limit of 2 to the select

mxm replied:
This doesn't work because we want to apply the `single_value` function iff the `SubqueryExpression` is **not** part of an array expression. Above, we would have to unwrap the single_value function again.

dobe commented on sql/src/main/java/io/crate/analyze/expressions/ExpressionAnalyzer.java:
```diff
@@ -848,11 +857,17 @@ protected Symbol visitSubqueryExpression(SubqueryExpression node, ExpressionAnal
+            }
+            // A SubQuery can return more than one row. We don't allow multiple rows,
+            // except in ANY or IN expressions. Thus, we wrap the result into a function
+            // which extracts the first element and checks if there are more than one element.
```
see comment above: creating an array is up to the array functions, this block could always return a selectsymbol.

mxm replied:
Yes, creating an array is but not ensuring that only one value is returned. Possibly we could return the select symbol and apply the `single_value` function at all places where we only want one value. The question is: Do we treat the `single_value` as the default or as a special case?
